mat64/cholesky.go

```diff
@@ -178,3 +178,155 @@ func (s *SymDense) InverseCholesky(chol *Cholesky) error {
+// returns whether the updated matrix A' is positive definite.
+//
+// The utility of RankOne comes from the fact that updating the factorization is
+// an O(n²) algorithm while computing it from scratch is an O(n³) algorithm.
+func (c *Cholesky) RankOne(orig *Cholesky, alpha float64, x *Vector) (ok bool) {
+	n := orig.Size()
+	if x.Len() != n {
+		panic(matrix.ErrShape)
+	}
+	if orig != c {
+		if c.chol == nil {
+			c.chol = NewTriDense(n, matrix.Upper, nil)
+		} else if c.chol.mat.N != n {
+			panic(matrix.ErrShape)
+		}
+		c.chol.Copy(orig.chol)
+	}
+
+	if alpha == 0 {
+		return true
+	}
+
+	// Algorithms for updating and downdating the Cholesky factorization are
+	// described, for example, in
+	// - J. J. Dongarra, J. R. Bunch, C. B. Moler, G. W. Stewart: LINPACK
+	//   Users' Guide. SIAM (1979), pages 10.10--10.14
+	// or
+	// - P. E. Gill, G. H. Golub, W. Murray, and M. A. Saunders: Methods for
+	//   modifying matrix factorizations. Mathematics of Computation 28(126)
+	//   (1974), Method C3 on page 521
+	//
+	// The implementation is based on LINPACK code
+	// http://www.netlib.org/linpack/dchud.f
+	// http://www.netlib.org/linpack/dchdd.f
+	// and
+	// https://icl.cs.utk.edu/lapack-forum/viewtopic.php?f=2&t=2646
+	//
+	// According to http://icl.cs.utk.edu/lapack-forum/archives/lapack/msg00301.html
+	// LINPACK is released under BSD license.
+	//
+	// See also:
+	// - M. A. Saunders: Large-scale Linear Programming Using the Cholesky
+	//   Factorization. Technical Report Stanford University (1972)
+	//   http://i.stanford.edu/pub/cstr/reports/cs/tr/72/252/CS-TR-72-252.pdf
+	// - Matthias Seeger: Low rank updates for the Cholesky decomposition.
+	//   EPFL Technical Report 161468 (2004)
+	//   http://infoscience.epfl.ch/record/161468
+
+	work := make([]float64, n)
+	blas64.Copy(n, x.RawVector(), blas64.Vector{1, work})
+
+	if alpha > 0 {
+		// Compute rank-1 update.
+
+		blas64.Scal(n, math.Sqrt(alpha), blas64.Vector{1, work})
+		umat := c.chol.mat.Data
+		for i := 0; i < n; i++ {
+			// Compute parameters of the Givens matrix that zeroes
+			// the i-th element of x.
+			c, s, r, _ := blas64.Rotg(umat[i*n+i], work[i])
+			if r < 0 {
+				// Multiply by -1 to have positive diagonal
+				// elemnts.
+				r *= -1
+				c *= -1
+				s *= -1
+			}
+			umat[i*n+i] = r
+			if i < n-1 {
+				// Multiply the extended factorization matrix by
+				// the Givens matrix from the left. Only
+				// the i-th row and x are modified.
+				blas64.Rot(n-i-1,
+					blas64.Vector{1, umat[i*n+i+1 : i*n+n]},
+					blas64.Vector{1, work[i+1 : n]},
+					c, s)
+			}
+		}
+		c.updateCond(-1)
+		return true
+	}
+
+	// Compute rank-1 downdate.
+
+	alpha = math.Sqrt(-alpha)
+	blas64.Scal(n, alpha, blas64.Vector{1, work})
+	// Solve U^T * p = x storing the result into work.
+	ok = lapack64.Trtrs(blas.Trans, c.chol.RawTriangular(), blas64.General{
+		Rows:   n,
+		Cols:   1,
+		Stride: 1,
+		Data:   work,
+	})
+	if !ok {
+		// The original matrix is singular.
```

Comment:
We get here if the factorization has zeros on the diagonal, that is, when the matrix was not positive definite. In this situation Factorize should have returned false. It seems to me that we don't say clearly what the user should do in such situation. It is also slightly related to #374.